clusterloader2/pkg/measurement/common/simple/service_creation_latency.go

```diff
@@ -0,0 +1,538 @@
+// "waitTillReady" waits until all services are available.
+// "gather" returns service creation latency summary.
+func (s *serviceCreationLatencyMeasurement) Execute(config *measurement.MeasurementConfig) ([]measurement.Summary, error) {
+	s.clusterFramework = config.ClusterFramework
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+
+	switch action {
+	case "start":
+		s.namespace, err = util.GetStringOrDefault(config.Params, "namespace", metav1.NamespaceAll)
+		if err != nil {
+			return nil, err
+		}
+		s.labelSelector, err = util.GetStringOrDefault(config.Params, "labelSelector", "")
+		if err != nil {
+			return nil, err
+		}
+		s.fieldSelector, err = util.GetStringOrDefault(config.Params, "fieldSelector", "")
+		if err != nil {
+			return nil, err
+		}
+		s.waitTimeout, err = util.GetDurationOrDefault(config.Params, "waitTimeout", defaultServiceCreationLatencyTimeout)
+		if err != nil {
+			return nil, err
+		}
+		return nil, s.start()
+	case "waitTillReady":
+		return nil, s.waitTillReady()
+	case "gather":
+		return s.gather()
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+// Dispose cleans up after the measurement.
+func (s *serviceCreationLatencyMeasurement) Dispose() {
+	if s.isRunning {
+		s.isRunning = false
+		close(s.stopCh)
+	}
+	s.queue.ShutDown()
+	s.workerGroup.Wait()
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	for _, pc := range s.pingCheckers {
+		pc.stop()
+	}
+	s.pingCheckers = make(map[string]*pingChecker)
+	if err := client.DeleteNamespace(s.clusterFramework.GetClientSets().GetClient(), s.probeNamespace); err != nil {
+		klog.Errorf("%s: namespace cleanup error: %v", s, err)
+	}
+}
+
+// String returns a string representation of the metric.
+func (*serviceCreationLatencyMeasurement) String() string {
+	return serviceCreationLatencyName
+}
+
+type serviceCreationTimes struct {
+	creation  time.Time
+	assigned  time.Time
+	available time.Time
+}
+
+func (s *serviceCreationLatencyMeasurement) start() error {
+	var err error
+	if err = s.setup(); err != nil {
+		return err
+	}
+	if s.informer != nil {
+		klog.Infof("%s: wait for controlled pods measurement already running", s)
+		return nil
+	}
+	klog.Infof("%s: starting wait for controlled pods measurement...", s)
+
+	s.isRunning = true
+	s.stopCh = make(chan struct{})
+
+	s.informer, err = s.newInformer()
+	if err != nil {
+		klog.Infof("%s: informer creation failed.", s)
+		return err
+	}
+
+	for i := 0; i < serviceCreationLatencyWorkers; i++ {
+		s.workerGroup.Start(s.worker)
+	}
+	go s.informer.Run(s.stopCh)
+	timeoutCh := make(chan struct{})
+	timeoutTimer := time.AfterFunc(informerSyncTimeout, func() {
+		close(timeoutCh)
+	})
+	defer timeoutTimer.Stop()
+	if !cache.WaitForCacheSync(timeoutCh, s.informer.HasSynced) {
+		return fmt.Errorf("timed out waiting for caches to sync")
+	}
+
+	return nil
+}
+
+func (s *serviceCreationLatencyMeasurement) setup() error {
```

Comment:
Personally, I wouldn't do that part of this measurement.
Similarly as we create prometheus stack, I would natively add to CL an ability to setup our "maintainance" namespace and start this pod there. [If we agree on that, please move it to a dedicate PR.]

@mm4tt - for thoughts

Reply:
I second Wojtek's idea, this "maintenance" namespace and pod(s) could be reused for other stuff in future. And this measurement may require the feature to be on, similarly to the ApiCallLatencyPrometheus measurement requiring Prometheus to be enabled.

Reply:
Done.